src/python/WMCore/MicroService/DataStructs/MSRuleCleanerWflow.py

```diff
@@ -70,7 +70,10 @@ def docSchema(self):
             ('ParentageResolved', True, bool),
             ('PlineMarkers', None, list),
             ('IsClean', False, bool),
-            ('ForceArchive', False, bool)]
+            ('IsLogDBClean', False, bool),
+            ('IsArchivalDelayExpired', False, bool),
```

Comment:
Given that `IsArchivalDelayExpired` depends solely on the request dictionary, I'm inclined to say that moving that code over to this module would make things more clear.

Reply:
At first glance I was about to agree with you immediately. But once I actually tried to implement it, I immediately ran into the need to import the `msConfig` object into `MSRuleCleanerWflow.py`, which makes things a little bit more complicate. And this made me immediately remember that we decided to keep this one as simple/generic as possible, but not putting a lot of weight to this class similar to `MSOutputTemplate.py`.

Reply:
I do agree with you that keeping as little as possible in the workflow representation object would be kind of a cleaner approach, but I am not quite sure how much of an impact would that have on the overall memory footprint from a single run. Because we are basically re-initializing the wflow object in place with every new workflow (see here [1]). And the memory should be release right after we get out of scope - which I expect to happen once the `_dispathWflow` function returns a value. And given the fact that we are feeding it a wflow at a time I thought the increased memory consumption while keeping the whole transition history of a single workflow would be neglectable compared to the enormous and detailed list of workflows returned by `reqmgr` and which we keep in memory until the end of the whole cycle (this is when this one gets out of scope). Of course all that  needs to be confirmed with a full run of a proper profiler, which I need to admit I haven't done - so there is a possibility that I may be wrong here. 

So I would suggest we  continue now as it have been already tested, and we may come back to this once we start thinking (eventually) about feeding the whole service in a different way. Will keep the conversation not resolve though - just for a future reminder.   

[1]
https://github.com/dmwm/WMCore/blob/875c3c5a48d0d82df947cdf13466c469c33ca27a/src/python/WMCore/MicroService/Unified/MSRuleCleaner.py#L188-L190

Reply:
And here is where you contradict yourself :-), compared to your previous comment:
"""
And this made me immediately remember that we decided to keep this one as simple/generic as possible, but not putting a lot of weight to this class similar to MSOutputTemplate.py.
"""

I try to push for workflow-related logic inside the workflow class, and the argument you give me is that you're concerned with the memory footprint...

The approach I would adopt is to add all the workflow logic in the same place, same as I've done for the workflow object used in MSTransferor:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/DataStructs/Workflow.py

Another good candidate for this would be a helper function that internally evaluates whether the workflow can be archived, e.g.:
```
if myWflowObj.canArchive():
    blah
```
and then that `canArchive` method would be implemented in this MSRuleWflow class and evaluate all the flags we need to evaluate.

Anyhow, I just wanted to make clear that my proposal was to inject workflow logic to the workflow struct class, not to make it as simple as possible and leave logic to be implemented elsewhere. As you said, this code works and it has been tested, let's move on!

Reply:
Agreed!